modules/platform/nuxeo-drive-server/nuxeo-drive-mongodb/src/main/java/org/nuxeo/drive/mongodb/MongoDBAuditChangeFinder.java

```diff
@@ -0,0 +1,195 @@
+
+        // Build intermediate filters
+        Bson lifeCycleEvent = and(eq("category", "eventLifeCycleCategory"), eq("eventId", "lifecycle_transition_event"),
+                ne("docLifeCycle", "deleted"));
+        Bson documentEvent = and(eq("category", "eventDocumentCategory"),
+                in("eventId", "documentCreated", "documentModified", "documentMoved", "documentCreatedByCopy",
+                        "documentRestored", "addedToCollection", "documentProxyPublished", "documentLocked",
+                        "documentUnlocked", "documentUntrashed"));
+        Bson inRootPath = regex("docPath", getRootsRegex(activeRoots.getPaths()));
+        Bson inRootId = in("docUUID", collectionSyncRootMemberIds);
+        Bson sessionRepository = eq("repositoryId", session.getRepositoryName());
+        Bson driveCategory = eq("category", "NuxeoDrive");
+        Bson notRootUnregistered = ne("eventId", "rootUnregistered");
+        Bson isGeneral = and(or(documentEvent, lifeCycleEvent), or(inRootPath, inRootId));
+        Bson isDrive = and(driveCategory, notRootUnregistered);
+        Bson idRange = and(gt("_id", lowerBound), lte("_id", upperBound));
+
+        // Build final filter
+        Bson filter = and(sessionRepository, or(isGeneral,
+                isDrive), idRange);
+
+        BsonDocument bsonDocument = filter.toBsonDocument(Document.class, auditCollection.getCodecRegistry());
+        log.debug("Query on MongoDB-Audit: {}", bsonDocument.toJson());
+
+        Bson order = Sorts.orderBy(Sorts.ascending("repositoryId"), Sorts.descending("eventDate"));
+        List<LogEntry> entries = auditCollection.find(filter)
+                                                .sort(order)
+                                                .map(MongoDBAuditEntryReader::read)
+                                                .into(new ArrayList<>());
+
+        // Post filter the output to remove (un)registration that are unrelated
+        // to the current user.
+        List<LogEntry> postFilteredEntries = new ArrayList<>();
+        String principalName = session.getPrincipal().getName();
+        for (LogEntry entry : entries) {
+            ExtendedInfo impactedUserInfo = entry.getExtendedInfos().get("impactedUserName");
+            if (impactedUserInfo != null && !principalName.equals(impactedUserInfo.getValue(String.class))) {
+                // ignore event that only impact other users
+                continue;
+            }
+            log.debug("Change detected: {}", entry);
+            postFilteredEntries.add(entry);
+        }
+        return postFilteredEntries;
+    }
+
+    protected String getRootsRegex(Set<String> roots) {
+        if(roots.isEmpty()) {
+            return "";
+        }
+        StringBuilder rootPathClause = new StringBuilder("^");
+        rootPathClause.append(String.join("|^", roots));
+        return rootPathClause.toString();
```

Comment:
```suggestion
        return "^" + String.join("|^", rootPathClause);
```
You don't need the `StringBuilder` here. As it is a simple append (no if or loop around), the compilator will produce more or less the same byte code.

Note: `+` is replaced by StringBuilder during compilation.

Reply:
Yes `+` is better. Note that in recent JDKs there are even better internal strategies than StringBuilder that are used to implement `+` on a fixed set of arguments.